pkg/planner/core/plan_cacheable_checker.go

```diff
@@ -658,7 +658,9 @@ func checkTableCacheable(ctx context.Context, sctx sessionctx.Context, schema in
 				return in, false // dynamic-mode for partition tables can use plan-cache
 			}
 		*/
-		return false, "query accesses partitioned tables is un-cacheable"
+		if sctx == nil || !sctx.GetSessionVars().IsDynamicPartitionPruneEnabled() {
```

Comment:
Should we introduce a `fix-control` as switch for safety? Then if there is any problem, the user can at least fall back to the previous behavior.
You can refer to this PR as an example: https://github.com/pingcap/tidb/pull/49754